pkg/lib/_internal/dnf5daemon.ts

```diff
@@ -0,0 +1,313 @@
+
+        if (pkgnames.length === 0)
+            return data;
+
+        async function refresh(session: string) {
+            // refresh dnf5daemon state
+            await call(session, "org.rpm.dnf.v0.Base", "read_all_repos", []);
+            const [, resolve_result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]) as [unknown[], number];
+            if (resolve_result !== 0)
+                throw new ResolveError(`Resolving read_all_repos failed with result={resolve_result}`);
+            await call(session, "org.rpm.dnf.v0.Goal", "do_transaction", [{}]);
+        }
+
+        async function resolve(session: string) {
+            const installed_names = new Set();
+            const seen_names = new Set();
+            const package_attrs = ["name", "version", "release", "arch", "download_size", "is_installed"];
+
+            const [results] = await call(session, "org.rpm.dnf.v0.rpm.Rpm", "list", [
+                {
+                    package_attrs: { t: 'as', v: package_attrs },
+                    scope: { t: 's', v: "all" },
+                    patterns: { t: 'as', v: pkgnames }
+                }
+            ]) as ListPackage[][];
+
+            for (const pkg of results) {
+                if (seen_names.has(pkg.name.v))
+                    continue;
+
+                if (pkg.is_installed.v) {
+                    installed_names.add(pkg.name.v);
+                } else {
+                    data.missing_ids.push(pkg.id.v);
+                    data.missing_names.push(pkg.name.v);
+                }
+
+                seen_names.add(pkg.name.v);
+            }
+
+            pkgnames.forEach((name: string) => {
+                if (!installed_names.has(name) && data.missing_names.indexOf(name) == -1)
+                    data.unavailable_names.push(name);
+            });
+        }
+
+        async function simulate(session: string) {
+            if (data.missing_ids.length === 0 || data.unavailable_names.length > 0) {
+                return null;
+            }
+
+            await call(session, "org.rpm.dnf.v0.rpm.Rpm", "install", [pkgnames, {}]);
+            const [transaction_items, result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]) as InstallResolveResult;
+            if (result !== 0)
+                throw new ResolveError(`Resolving install failed with result={result}`);
+            for (const transaction_item of transaction_items) {
+                const [sort, _action, reason, _leftover, pkg] = transaction_item;
+                const name = pkg.name.v;
+
+                if (sort !== "Package")
+                    continue;
+
+                data.download_size += pkg.download_size.v;
+
+                if (reason == "Dependency") {
+                    data.install_ids.push(pkg.id.v);
+                    if (data.missing_names.indexOf(name) == -1)
+                        data.extra_names.push(name);
+                }
+
+                // TODO: support removals
```

Comment:
Not sure what the `reason` or `action` would be in this case. Is there an example of a package replacing another one in Fedora.